### PR TITLE
Add size limit to tmpfs

### LINKED ISSUE #63126: emptyDir with medium: Memory ignores sizeLimit
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

The tmpfs volume that's created with emptyDir doesn't have the correct size limit. In fact, the tmpfs volume size is always half of the available RAM on the Kubernetes node, which is the default behavior on a Linux kernel (https://www.kernel.org/doc/Documentation/filesystems/tmpfs.txt). This implies that the size option was likely not provided when mounting a tmpfs. This is similar to the issue here: https://github.com/kubernetes/kubernetes/issues/48912.

Note: All the Kubernetes components have the necessary feature gates enabled.

**What you expected to happen**:

The tmpfs volume should respect the `sizeLimit` parameter, unless the `sizeLimit` parameter is not meant for tmpfs.

**How to reproduce it (as minimally and precisely as possible)**:

Step 1: Create a deployment that with an emptyDir volume with `medium` set to Memory and `sizeLimit` set to 1Gi.
```
$ cat sample.yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
spec:
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - image: nginx
        name: nginx
        volumeMounts:
          - mountPath: /scratch
            name: scratch
      volumes:
      - name: scratch
        emptyDir:
          medium: Memory
          sizeLimit: "1Gi"

$ kubectl apply -f sample.yaml
deployment "test" configured
```

Step 2: Inspect the volume mount and the size is not 1GB. Instead, the size is 126GB, because the Kubernetes node has 256GB of available RAM.
```
kubectl exec -it test-56f7bb87d5-j68n5 bash
root@test-56f7bb87d5-j68n5:/# mount -l | grep scratch
tmpfs on /scratch type tmpfs (rw,relatime)
root@test-56f7bb87d5-j68n5:/# df -h /scratch
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           126G     0  126G   0% /scratch
```

Step 3: When exporting the deployment manifest file, the `sizeLimit` option is retained, demonstrating that the necessary feature gates were enabled.

```
kubectl get pods test-56f7bb87d5-j68n5 -o yaml --export
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  generateName: test-56f7bb87d5-
  labels:
    app: test
    pod-template-hash: "1293664381"
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: test-56f7bb87d5
    uid: 63b2a2d3-4815-11e8-aacf-ec0d9a749964
  selfLink: /api/v1/namespaces/default/pods/test-56f7bb87d5-j68n5
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /scratch
      name: scratch
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-qb9lt
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: <HOST>
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir:
      medium: Memory
      sizeLimit: 1Gi
    name: scratch
  - name: default-token-qb9lt
    secret:
      defaultMode: 420
      secretName: default-token-qb9lt
status:
  phase: Pending
  qosClass: BestEffort
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:21:50Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:13:31Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:

The api server, controller manager, scheduler, and kubelets have the following feature gates enabled:
```--feature-gates=PersistentLocalVolumes=true,VolumeScheduling=true,MountPropagation=true,LocalStorageCapacityIsolation=true,DevicePlugins=true```

The only feature gate that is needed is `LocalStorageCapacityIsolation`, but the other features are enabled for other reasons.

The Kubernetes nodes in this cluster have 256GB of RAM.
```
$ free -g
              total        used        free      shared  buff/cache   available
Mem:            251           2         238           2          10         245
```

- OS (e.g. from /etc/os-release):

```
 cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.4 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.4 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

- Kernel (e.g. `uname -a`):

```
uname -a
Linux <host> 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

- Install tools:

`kubeadm`.

- Others:

### PR DESCRIPTION
**Pull request by lovejoy**
**What this PR does / why we need it**:
Add a size option when mount tmpfs 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63126

**Special notes for your reviewer**:
i run a pod like below 
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox-1
  namespace: default
spec:
  containers:
  - command:
    - sleep
    - "360000"
    image: busybox
    imagePullPolicy: IfNotPresent
    name: busybox
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
      - name: foo
        mountPath: /data/mysql
    resources:
      limits:
        memory: 1229Mi
      requests:
        cpu: 500m
        memory: 1Gi
  volumes:
   - name: foo
     emptyDir:
       sizeLimit: "350Mi"
       medium: "Memory"
  dnsPolicy: ClusterFirst
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  terminationGracePeriodSeconds: 30
```
and it  have a sizelimit now 
/ # df -h | grep mysql
tmpfs                   350.0M    350.0M         0 100% /data/mysql
**Release note**:

```release-note
Fixed emptyDir to use the sizeLimit when creating the temporary volumes. Before this fix the tmpfs volume size was set to half of the available RAM (default linux kernel behavior)
```

### CODE REVIEW

**Comment by lovejoy** [comment]
/assign @thockin
cc @jingxu97

File: pkg/volume/empty_dir/empty_dir.go [hunk 1]
```diff
@@ -107,15 +107,18 @@ func (plugin *emptyDirPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, opts vo
 
 func (plugin *emptyDirPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, mounter mount.Interface, mountDetector mountDetector, opts volume.VolumeOptions) (volume.Mounter, error) {
 	medium := v1.StorageMediumDefault
+	sizeLimit := &resource.Quantity{}
 
 	if spec.Volume.EmptyDir != nil { // Support a non-specified source as EmptyDir.
 		medium = spec.Volume.EmptyDir.Medium
+		sizeLimit = spec.Volume.EmptyDir.SizeLimit
```
Comment (dims) [inline pkg/volume/empty_dir/empty_dir.go]: need to check for nil? EmptyDirVolumeSource's SizeLimit is `*resource.Quantity`

File: pkg/volume/empty_dir/empty_dir.go [hunk 2]
```diff
@@ -253,8 +258,14 @@ func (ed *emptyDir) setupTmpfs(dir string) error {
...
 
-	glog.V(3).Infof("pod %v: mounting tmpfs for volume %v", ed.pod.UID, ed.volName)
-	return ed.mounter.Mount("tmpfs", dir, "tmpfs", nil /* options */)
+	var options []string
+	// Linux system default is 50% of capacity.
+	if ed.sizeLimit != nil {
+		options = []string{fmt.Sprintf("size=%d", ed.sizeLimit.Value())}
+	}
+	glog.V(3).Infof("sizeLimit : %#v", ed.sizeLimit)
+	glog.V(3).Infof("pod %v: mounting tmpfs for volume %v with options $v", ed.pod.UID, ed.volName, options)
```
Comment (dims) [inline pkg/volume/empty_dir/empty_dir.go]: `options` has the sizeLimit from previous line, so we don't need line 266. right?

File: pkg/volume/empty_dir/empty_dir.go (hunk 1 above)
Comment (lovejoy) [reply pkg/volume/empty_dir/empty_dir.go]: ok

**Comment by lovejoy** [comment]
i already delete it in after commit

File: pkg/volume/empty_dir/empty_dir.go (hunk 2 above)
Comment (lovejoy) [reply pkg/volume/empty_dir/empty_dir.go]: i already delete it in after commit

Comment (dims) [reply pkg/volume/empty_dir/empty_dir.go]: oops. thanks!. Can you please squash the two commits into one?

Comment (sjenning) [reply pkg/volume/empty_dir/empty_dir.go]: options $v?  should be %v

Comment (lovejoy) [reply pkg/volume/empty_dir/empty_dir.go]: orz... i will fix this

**Comment by dims** [comment]
Please add a release note

**Comment by lovejoy** [comment]
@dims added the release note and may be you have some better describe for the release note？

**Comment by dims** [comment]
How about this?

`Fixed emptyDir to use the sizeLimit when creating the temporary volumes. Before this fix the tmpfs volume size was set to half of the available RAM (default linux kernel behavior)`

**Comment by msau42** [comment]
We need to carefully consider and document the behavior implications of this.  The emptyDir sizeLimit for non-tmpfs volumes is a soft limit, ie pods get evicted vs IO error in the app.  But this change makes it a hard limit for tmpfs.  So users are going to see different behavior of sizeLimit depending on the emptydir type, and that could be confusing.

I think this same difference exists when handing limits for cpu/memory, so it could just be a matter of good documentation.

**Comment by lovejoy** [comment]
@msau42 ok i will update the document about tmpdir in the website

**Comment by jingxu97** [comment]
I want to double check the use case of this PR. Currently, for emptyDir backed by memory, if the usage exceeds limit, the pod will be gracefully terminated and rescheduled. After this PR, the memory is a hard limit, if the emptyDir usage exceeds the limit, applications might get errors instead of evicted. Is this the desired behavior in what use cases?

**Comment by thockin** [comment]
I hate to be the naysayer, but the way this is presented through the API implies that it has meaning regardless of `medium`.  But it doesn't.  It is totally ignored for both `medium: ""` and `medium: HugePages`.  We either need to:

a) Implement it for those (not sure how to do it for `""` except by something like grpquota which requires a supplemental GID to be added per volume)

b) Error on validation (ugly)

c) Change the API so it applies to just memory.  e.g.

```
volumes:
  - name: tmpdata
    emptyDir:
      medium: Memory
      memory:
        limit: 1GiB
```

I prefer (c), unless we can make (a) work.

Jing's concern is legit - probably should validate that during API validation.

We must also document that this comes out of the pod's overall memory limit (right?).

At some point there was discussion about how we might limit this in alignment with the ephemeral storage proposals - @jingxu97  ?

**Comment by thockin** [comment]
Sorry, some of what I said was already merged, and I just forgot :)

Some of it is still valid.

**Comment by derekwaynecarr** [comment]
this requires some special review - especially for when the /tmpfs is reclaimed in pod lifecycle.  ultimately usage is capped by memory cgroup.  like tim, i need some time to reflect on impact of this.

**Comment by derekwaynecarr** [comment]
@jingxu97 i think we should discuss this some in sig-node meeting this week.

**Comment by thockin** [comment]
FWIW I think it should be fixed, but we need to make sure we understand HOW
and WHY :)

**Comment by lovejoy** [comment]
@thockin i hope there's  a solution can make a work. we exactly need hard limit for all kinds emptyDir because user may use out all the space accidently ,this affects the whole node and 
 is really  happning
 in our production  environment.

**Comment by jingxu97** [comment]
@derekwaynecarr sure, we can discuss more in sig-node. But I was in f2f meeting this Tuesday. Is that ok for next sig-node meeting? Thanks!

**Comment by lovejoy** [comment]
any update for this ?

**Comment by 2rs2ts** [comment]
Just my $0.02 as a user but I find it _very_ surprising that, when I violate a size limit, instead of getting an error that my code can handle I just get my container shut down (albeit gracefully.) It's definitely very important to ruminate on the impact this change could cause, and possibly change the API so people don't get suddenly surprised by the change, but you should definitely let users actually know that they violated the size limit when they do, so they can work around it rather than just getting evicted and then violating the limit on another node, etc. etc. ;)

**Comment by vishh** [comment]
If memory limits for tmpfs becomes an explicit config knob, what would the overall pod memory limit be? Imagine a pod with tmpfs limit of `2Gi`, and a container with memory limit `1Gi`, then will the pod limit be `3Gi`? The container will still get OOM killed if it writes more than `1Gi` to it's tmpfs. If the pod limit is `1Gi` then the tmpfs limit doesn't make any sense except in the case of container restarts (which is a corner case).

For this PR to make sense, we need kernel support to charge tmpfs usage to the pod level instead of container/first-touch level.

**Comment by mikkeloscar** [comment]
I think having a tmpfs volume contribute to the same limit as pod memory is what makes it hard to use.

Ideally you would have a limit on the tmpfs which would give you out of space errors if you try to write more data than the limit, but it would not contribute to the memory limit set on the containers in the pod. Obviously you need to take all limits (or requests really) into account when scheduling the pod.
The fact that tmpfs is using memory is just an implementation detail in my mind, it's still a volume and other volume types are treated as a separate resource from the pod (for obvious reasons), but this makes it easy to reason about how much you can write to your volume vs. how much memory your process can consume and you don't have to worry that these concerns/limits might overlap.

Whether it's feasible to implement it like this with the way pods are implemented, I'm not sure?

**Comment by lovejoy** [comment]
@vishh I think what you describe won't cause oom ,It just will be a disk full error. I will test this later.

**Comment by thockin** [comment]
I am very interested in this PR, but since it is idle, I am unassigning myself.  PLEASE add me back when you're ready to proceed.

**Comment by warmchang** [comment]
https://github.com/kubernetes/kubernetes/pull/63641#issuecomment-405003281
https://github.com/kubernetes/kubernetes/pull/63641#issuecomment-404924118

To be honest, I have the same confusion.

@lovejoy , Is there any test conclusion to share now? Thanks.

**Comment by lovejoy** [comment]
@warmchang   @vishh  @mikkeloscar 

I did a test : a pod with tmpfs limit of 200M and memory limit of 100M, and  write 150M file to empty_dir like below and the pod didn't oom
```
~ # df -h | grep mysql
tmpfs                   200.0M    150.0M     50.0M  75% /data/mysql
~ # cat /sys/fs/cgroup/memory/memory.limit_in_bytes
104857600
```


docker  stats output
```
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O
c84ab605d9a9        0.00%               1.061 MB / 104.9 MB   1.01%               0 B / 0 B           89.46 MB / 46.74 MB
```

Note: Theres still something confusing me. If I `dd of=/dev/zero of=1.img` in container ,the dd command will be killed by oom killer. But if I docker cp a big file to  empty_dir ,It's all right, and the file content is correct I can tail or head to check it as less will be killed .

And I think the empty_dir is just a HostPath bind to docker like we docker run -v xxx:xxx 
```
      "HostConfig": {
            "Binds": [
                "/home/kubelet/pods/b5729eaa-bcb8-11e8-8aed-e0db55136a58/volumes/kubernetes.io~empty-dir/foo:/data/mysql",
```
It would not be  caculated by cgroup ?

**Comment by vishh** [comment]
@lovejoy the memory occupied by the tmpfs files will be charged to the memory cgroup of the writing process. In the case of `dd` from within the container, the `dd` process is in the memcg of the container. I don't remember `docker cp` implementation anymore and so I'm not sure which process and which memcg is writing files to the tmpfs accessible inside the container.

@adityakali what would it take to upstream patches to specify memcg per tmpfs?

**Comment by adityakali** [comment]
> @adityakali what would it take to upstream patches to specify memcg per tmpfs?
Can you please open a FR for this for COS?

Looping in @gthelen who may have additional input on whether that patch will really benefit this particular use-case. 

Reading through the bug, it looks like the memory is already being charged to the writer's memcg in case of k8s. So at least setting `memcg=<writer's memcg>` seems redundant.

**Comment by vishh** [comment]
@adityakali The intention as I understand it is to be able to charge usage of tmpfs to a k8s pod instead of the containers that are consuming it.

**Comment by thockin** [comment]
This is back in my queue.  We know that tmpfs charging is not exactly what we want here, but is it workable?  Rather, is this any worse than no limit?  I don't think it is.

Ergo, I am inclined to approve, but I'd like a node from @vishh and @msau42 that this is unlikely to explode on us

**Comment by thockin** [comment]
Actually, I am confused.  Docs for `EmptyDirVolumeSource` say:

```
    // Total amount of local storage required for this EmptyDir volume.
    // The size limit is also applicable for memory medium.
    // The maximum usage on memory medium EmptyDir would be the minimum value between
    // the SizeLimit specified here and the sum of memory limits of all containers in a pod.                                                                                             
    // The default is nil which means that the limit is undefined.
    // More info: http://kubernetes.io/docs/user-guide/volumes#emptydir
    // +optional
    SizeLimit *resource.Quantity
```

This was written by @jingxu97 in commit 85f030c2aad00c54d498646e4151e7fad930c7c3 but that commit does not seem to have any impl for it..

So what is going on???

**Comment by jingxu97** [comment]
@thockin, this PR https://github.com/kubernetes/kubernetes/pull/45686 is the code to implement the sizeLimit for emptyDir. In the design, the sizeLimit parameter set for emptyDir volume could not be used for creating a volume with the size. Instead what it does it that eviction manager keeps monitoring the disk space used by pod emptyDir volume and it will evict pods when the usage exceeds the limit.

**Comment by vishh** [comment]
We preferred soft limits (based on background eviction loops) instead of hard limits (setting the limit on tmpfs to be that emptydir size) to allow for kubelet to detect and expose situations where the pod is actually running out of space on it's empty dir volume. This lets users (or automation systems like vertical pod autoscaling) bump up limits on-demand. If we instead set the limit to `sizeLimit`, the app would receive a ENOSPC error (which is probably expected from the developer POV), which is often not exposed outside of apps clearly or could be attributed to running out of other resources on the node like pids for example. 
In the case of tmpfs, the memory limit of the pod acts as a safety net. 
If hard limits for tmpfs is needed, I'd prefer that being a non-default option controllable via k'let configuration. I can see it being necessary in scenarios where there is a high risk of DDOS, but I'd prefer keeping the default user experience simpler.

**Comment by mikkeloscar** [comment]
> I can see it being necessary in scenarios where there is a high risk of DDOS, but I'd prefer keeping the default user experience simpler.

@vishh I don't follow your argument for how this is simpler without hard limits? I think it depends on what problem you are trying to solve. Let me share a use case that this change would help me solve.

We have an Elasticsearch cluster where we want to run some of the "nodes" with tmpfs storage for fast read performance. In an environment like AWS EC2 we would simply allocate x% of the instance memory as storage space and the other y% of the instance memory for the Elasticsearch application.
If we want to do this with Kubernetes we can currently only use 50% of the nodes memory for tmpfs because this is how much is allocated when you define an emptyDir of medium `Memory` and we have a pod that is less than 50% we can not easily prevent the space used for "storage" to overlap with that of the application.

Additionally we can not just allocate a tmpfs manually and mount it e.g. as a `hostPath` because of how disk caches are allocated to the total memory of the `cgroup` of a pod/container as far as I understand. That is, if we allocate say 60% of a nodes memory to a tmpfs and the rest to a pod/container then the container could get OOM Killed when writing to the tmpfs even though the process itself didn't use 40% of memory. At least this is what we have observed.
In this scenario we could of cause run the pod/container without memory limits, but this only really works (and not in practice) if there is nothing else running on that same node, otherwise we might just kill some other random application when starting to use more memory than requested.

Do you have an idea for how to solve this problem, either with this change, or some other way?

**Comment by vishh** [comment]
Appreciate the additional context. Firstly, I assume you have set `node
allocatable` to say `10 %` of your instance memory to protect your node
from system OOMs. (In reality `10%` may not be the optimal value and you
need to measure overhead for your clusters).

I have a few more questions to better understand your situation.
Let's say you have a 4 Gi VM, and you are allocating 3Gi for your elastic
search pod. How many containers are there in the ES pod? Do they all have
memory limits? If no, why not? If yes, are you running into a situation
where a side-car container in the ES pod gets killed by the kernel OOM
killer because of the ES container consuming more tmpfs (via emptydir)
space than expected? If so, is there no way to configure ES to limit the
amount of storage it would consume?

**Comment by mcluseau** [comment]
isn't a solution to this in a new parameter `hardSizeLimit`?

**Comment by sftim** [comment]
In terms of tmpfs emptyDir volumes and memory cgroups, it seems that it's handy to have an option to charge the tmpfs memory to the Pod (or to a cgroup inside the Pod) rather than the container.

Eg: An emptyDir volume in a pod with 2 containers. Containers A and B have a 512 MiB memory limit.

Container A has a fairly small process that writes 499 MiB of app data to the tmpfs and doesn't OOM whilst doing so. Container B has a process twice the size of container A's that rewrites the file with updated app data. Container B OOMs because writing those pages took ownership of them, and the process plus the app data totals more tha 512MiB.

This feels like it'd surprise the developer writing code to run in those containers, who'd expect to be able to set something like:
```yaml
      requests:
        memory: "4Mi"
      limits:
        memory: "16Mi"
```
for both containers (and if the Volume were a different type, that'd work fine).

**Comment by derekwaynecarr** [comment]
this had fallen into the back of my queue as well.

if a container writes to an emptyDir, and the container restarts, the charge is transferred to the bounding pod cgroup. this makes me wonder if its better to default the size of the tmpfs to node allocatable.  basically, why treat tmpfs differently?  if this was the default behavior, would you still desire the ability to specify a more granular limit?

**Comment by dims** [comment]
@lovejoy can you please answer the last question from @derekwaynecarr (and any other feedback as well)

**Comment by xordspar0** [comment]
> this makes me wonder if its better to default the size of the tmpfs to node allocatable.

@derekwaynecarr Could you explain more about what you mean by this? Is "node allocatable" the space total space on a node available to all pods? Currently the tmpfs size is half of the node's total memory, correct?
Can you explain what this would do to improve the situation?

**Comment by bashimao** [comment]
@lovejoy FYI, https://github.com/kubernetes/kubernetes/issues/28272#issuecomment-549223540

**Comment by dims** [comment]
/milestone v1.18

let's work on this for the next release please. ( cc @alejandrox1 @krmayankk )

**Comment by annajung** [comment]
Hello @lovejoy @msau42 @sjenning @dims   👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this issue is tagged for 1.18, I'd like to check its status. Is this issue intended for milestone 1.18?

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by annajung** [comment]
Hello @lovejoy @msau42 @sjenning @dims Checking with you all again. Is this issue intended for milestone 1.18?

**Comment by dims** [comment]
@annajung looks like this did not progress in a while ..

**Comment by scrummyin** [comment]
After reading the conversation on this I think the part where there seems to be a hangup is around the goal of emptyDir? Right now it just points to `mount(tmpfs, ...)`. So, from what I can tell the goal is to map to `tmpfs`, which results in a fast and ephemeral directory. Which means this PR makes a lot of sense since it exposes one more mount option of `tmpfs`, and paves the way for more options to be added. Which allows informed users more flexibility and does not hinder current users since it defaults to the previous value of `nil`.

The last bit of conversation seems to be hung up on if there is a better way. If there is a better way this PR is still a step in getting there. Since allowing the ability to set *any* value, is a step towards setting *the best* value.

**Comment by jonaz** [comment]
Why is this not merged? We want to run our jenkins builds in kubernetes with emptydir memory but cannot do that right now until this is merged.

**Comment by wesleyw72** [comment]
+1 to the above comments. @dims sorry to tag you specifically, but you're one of the most active thus far. What are the next steps for getting this in?

**Comment by dims** [comment]
/assign @xing-yang 

@msau42 @xing-yang this PR has been languishing for some time, can you please make a call one way or another?

/milestone v1.19

**Comment by sftim** [comment]
Suggestion for release note:
```
Fixed emptyDir to use the sizeLimit when creating temporary volumes. Before this fix, the tmpfs volume size relied on default kernel behavior (Linux: half of the available RAM).
```

@lovejoy what do you think?

**Comment by cvanlabe** [comment]
Just to confirm, this would _not_ be affected by memory resource limits right? 

Currently, it's not entirely clear to me from above conversations, current experience, and the statement _"Memory-backed volume sizes are now bounded by cumulative pod memory limits."_ on https://bugzilla.redhat.com/show_bug.cgi?id=1422049 (openshift I know) if the memory limit set also needs to incorporate the memory volume or not.

ie: some say or give the impression the memory limit kills a pod once the memory backed volume goes over the said limit, other discussion outputs indicate there's no problem and is tracked separately.

**Comment by scrummyin** [comment]
I think based off of section `2.3 Shared Page Accounting` and `10. OOM Control` in https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt it might depend on your specific situation, but this PR should not change the current behavior around memory resource limits. Instead, it allows for more customization over resource usage. In the [bug you cited](https://bugzilla.redhat.com/show_bug.cgi?id=1422049), this feature is listed as one of the expected result, which I read as a possible solution to it. In theory, tmpfs size limits could be used to mitigate the issue of a container running out of memory by allowing users to set a smaller tmpfs size so that they would first hit the tmpfs limit before running out of memory.

**Comment by cvanlabe** [comment]
Thanks Brian. I don't want to sidetrack this conversation, but would really like to understand how someone can actually set a guaranteed memory volume size (if this is possible at all). As a practical example, I have a node with 240G memory, and tmpfs got half of that so 120G. The problem is that a lot of things use that tmpfs... secrets are stored there, CNIs like Calico use it, and others. I run an application which expects to have a certain amount of disk space, now a memory filesystem, but had a few occurrences where we ran out of disk space, because apparently, we have less than 120G, but hard to say how much exactly. 

Is there a way to do the opposite of setting a sizeLimit, but more of a sizeRequest? (Can open a new github issue for this is this is too much distracting..)

**Comment by msau42** [comment]
/assign @derekwaynecarr @dashpole 
to review. This change has implications on memory accounting so I think sig-node would be best to drive the direction of this.

**Comment by dashpole** [comment]
I've read through the history, and think there are two separate problems.

1. The default value for the size limit breaks some applications
2. We want to add a feature to be able to limit the size of tmpfs

Based on the discussion above, users mostly care about 1, which sounds like a bug.  I don't see anyone that actually wants 2, which would require some consideration to get right, and which we should implement for all medium types.

I'm in favor of @derekwaynecarr's [suggestion above](https://github.com/kubernetes/kubernetes/pull/63641#issuecomment-509391492) to set the tmpfs limit to allocatable.  Alternatively, we could set the tmpfs limit to the pod's memory limit.  Both of these seem preferable to the current behavior, and don't have any downsides I can think of.

**Comment by 2rs2ts** [comment]
>we could set the tmpfs limit to the pod's memory limit
This please. Or perhaps its request. The reason why is so that the scheduler can take it into account.

**Comment by cvanlabe** [comment]
For my use case I do not prefer the pods memory limit. I actually want to make sure the pod won't consume more than a specific amount of memory before being killed and see that independent of its (memory backed) volumes that may and are expected to grow to much higher sizes.

Eg: server with 240G.. I want 20G for the OS doing it's base functionality, 20G for the pod. A healthy run wouldn't consume more than that. 200G for memory backed fast filesystem.

If it's depending on the pods memory limit, I would have to set it to 220G, but when my old process would get out of control and grows to 50G say, my filesystem would not be my expected 200G anymore and a lot more unexpected things can happen.

sizeLimit and sizeRequest would be my preference.

**Comment by wangjia184** [comment]
When can this fix be released?

**Comment by JoshBroomberg** [comment]
@dims any progress on this? We're looking to limit the size of the tmpfs that we're using to increase the size of the shared memory available to pods. We risk pod eviction without this kind of limit.

**Comment by matti** [comment]
@JoshBroomberg can tou describe how you are using tmpfs on your pods? I think that this thread needs more concrete use cases.

**Comment by JoshBroomberg** [comment]
@matti we run user (long-living) jobs based on programmatically generated pod specs. Some of these user jobs involve things like training DL models using PyTorch multiprocessing. This often requires more than 64mb shared memory.

Our temporary work around is to mount a memory-backed empty dir to `/dev/shm` (as is done by many others). The annpying thing about this is that, for all other memory use, we use the pod spec for careful control. IE, users are able to select, and be limited to, an exact amount of memory. But we we use the tmpfs workaround, we violate this limit. Pods can use up to half of the node memory. 

We haven't released this option yet because we are concerned about the side effects of this usage and the resulting UX. We will release without the limit (because this is badly needed by some). But we would love the kind of limit created by this PR.

**Comment by derekwaynecarr** [comment]
The proposal from @dashpole is the path I see as safe to move forward.
https://github.com/kubernetes/kubernetes/pull/63641#issuecomment-655726138

If all containers in a pod specify a memory limit, we could set to limit of pod.
If all containers in a pod do not specify a memory limit, we could set to node allocatable.
Its important to remember that when a container restarts, the charge for writes to that emptyDir transfer back to the pod level cgroup when doing memory accounting.

**Comment by derekwaynecarr** [comment]
I am putting a PR together that combines a user specifiable setting with a feature gate to control rollout.
Will link shortly.

**Comment by derekwaynecarr** [comment]
see: https://github.com/kubernetes/kubernetes/pull/94444

**Comment by mitar** [comment]
Shouldn't the issue be closed once PR is merged?

**Comment by derekwaynecarr** [comment]
@mitar see linked pr and kep.
https://github.com/kubernetes/kubernetes/pull/94444
https://github.com/kubernetes/enhancements/pull/1968

**Comment by mitar** [comment]
Yes, I have seen the PR, but PR is not yet merged.

**Comment by 2rs2ts** [comment]
@mitar this is a PR you're commenting on. The parent issue is still open https://github.com/kubernetes/kubernetes/issues/63126

**Pull request closed without merging**